src/main/java/com/android/volley/toolbox/CacheHeader.java

```diff
@@ -115,6 +120,41 @@ static CacheHeader readHeader(DiskBasedCache.CountingInputStream is) throws IOEx
                 key, etag, serverDate, lastModified, ttl, softTtl, allResponseHeaders);
     }
 
+    /**
+     * Reads the header from a AsynchronousFileChannel and returns a CacheHeader object.
+     *
+     * @param afc File channel to read from.
+     * @param size Size of the cache header.
+     * @throws IOException if fails to read header
+     */
+    @RequiresApi(api = Build.VERSION_CODES.O)
+    static CacheHeader readHeader(final AsynchronousFileChannel afc, int size) throws IOException {
+        final ByteBuffer buffer = ByteBuffer.allocate(size);
+        Future<Integer> fillBuffer = afc.read(buffer, 0);
+        try {
+            fillBuffer.get();
+            buffer.flip();
+            int magic = buffer.getInt();
+            if (magic != CACHE_MAGIC) {
+                // don't bother deleting, it'll get pruned eventually
+                throw new IOException();
+            }
+            String key = DiskBasedCacheUtility.readString(buffer);
+            String etag = DiskBasedCacheUtility.readString(buffer);
+            long serverDate = buffer.getLong();
+            long lastModified = buffer.getLong();
+            long ttl = buffer.getLong();
+            long softTtl = buffer.getLong();
+            List<Header> allResponseHeaders = DiskBasedCacheUtility.readHeaderList(buffer);
+            return new CacheHeader(
+                    key, etag, serverDate, lastModified, ttl, softTtl, allResponseHeaders);
+        } catch (ExecutionException e) {
```

Comment:
Be careful about reporting execution exceptions, which are not expected, as ioexceptions.

Reply:
Do you suggest adding these exceptions to the method header, then catching them like the IOException in DiskBasedAsyncCache? I don't see how they'd be handled differently, since we still would have to return null on the get call or skip the file in initialize.

Reply:
It is tricky mixing sync and async code. Java doesn't have a great solution yet AFAIK. We punted on this when we created the Executor framework in java.util.concurrent.

1. Avoid morphing non-IO exceptions into IO exceptions. IOExceptions are checked and have specific causes. In this scenario, there shouldn't be any other checked exceptions, right? Except possibly interrupted exception, but it can be rethrown. Any other exception is probably a programming error and should not be disguised, caught, and then handled by the client's IOException handler.

2. Consider mimicking what would happen if the code executed synchronously. If an unchecked exception is the cause you could attempt to rethrown it in some fashion.

There are a few solutions floating around. Some of them are attempts to avoid checked exceptions altogether, but this is a more noble endeavor.

Reply:
I decided to return null instead of throwing it as an IOException, then dealing with it in DBAC. Let me know if there are any issues with this approach.